From de404aaa671e8b05603ba04d30dd6d1440c3f5f3 Mon Sep 17 00:00:00 2001
From: Zidan Wang <zidan.wang@freescale.com>
Date: Tue, 31 Mar 2015 21:03:53 +0800
Subject: [PATCH 0521/1691] MLK-10108-1: ASoC: fsl_sai: Add tdm slot operation
 for SAI master mode

commit de404aaa671e8b05603ba04d30dd6d1440c3f5f3 from
git://git.freescale.com/imx/linux-2.6-imx.git imx_4.1.15_1.0.0_ga

Add tdm slot operation for SAI master mode. When using SAI as master
mode, we should use set_tdm_slot() helper function to set tdm slots in
machine driver.

SAI will generate BCLK depends on sample rate, slots and slot width.
And there may be unused BCLK cycles before each LRCLK transition.

Signed-off-by: Zidan Wang <zidan.wang@freescale.com>
(cherry picked from commit 06fe60409a2de3192cee78331fcb8baca337db69)
---
 sound/soc/fsl/fsl_sai.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/sound/soc/fsl/fsl_sai.c b/sound/soc/fsl/fsl_sai.c
index 42eb5d6..27478ef 100644
--- a/sound/soc/fsl/fsl_sai.c
+++ b/sound/soc/fsl/fsl_sai.c
@@ -394,11 +394,14 @@ static int fsl_sai_hw_params(struct snd_pcm_substream *substream,
 	unsigned int channels = params_channels(params);
 	u32 word_width = snd_pcm_format_width(params_format(params));
 	u32 val_cr4 = 0, val_cr5 = 0;
+	u32 slot_width = word_width;
 	int ret;
 
 	if (!sai->is_slave_mode) {
+		slot_width = sai->slot_width;
 		ret = fsl_sai_set_bclk(cpu_dai, tx,
-				sai->slots * word_width * params_rate(params));
+				sai->slots * slot_width * params_rate(params));
+
 		if (ret)
 			return ret;
 
@@ -410,14 +413,13 @@ static int fsl_sai_hw_params(struct snd_pcm_substream *substream,
 
 			sai->mclk_streams |= BIT(substream->stream);
 		}
-
 	}
 
 	if (!sai->is_dsp_mode)
-		val_cr4 |= FSL_SAI_CR4_SYWD(word_width);
+		val_cr4 |= FSL_SAI_CR4_SYWD(slot_width);
 
-	val_cr5 |= FSL_SAI_CR5_WNW(word_width);
-	val_cr5 |= FSL_SAI_CR5_W0W(word_width);
+	val_cr5 |= FSL_SAI_CR5_WNW(slot_width);
+	val_cr5 |= FSL_SAI_CR5_W0W(slot_width);
 
 	if (sai->is_lsb_first)
 		val_cr5 |= FSL_SAI_CR5_FBT(0);
@@ -769,6 +771,7 @@ static int fsl_sai_probe(struct platform_device *pdev)
 	}
 
 	sai->slots = 2;
+	sai->slot_width = 32;
 
 	irq = platform_get_irq(pdev, 0);
 	if (irq < 0) {
-- 
1.9.1

